lib/rex/proto/dns/resolver.rb

```diff
@@ -132,8 +132,8 @@ def send(argument, type = Dnsruby::Types::A, cls = Dnsruby::Classes::IN)
       end
 
       # Store packet_data for performance improvements,
-      # so methods don't keep on calling Packet#data
-      packet_data = packet.data
+      # so methods don't keep on calling Packet#encode
+      packet_data = packet.encode
```

Comment:
This unfortunately introduces a bug in the `auxiliary/gather/enum_dns` module which I'm assuming uses a different object type for `packet`.

```
msf6 auxiliary(gather/enum_dns) > run

[-] Auxiliary failed: NoMethodError undefined method `encode' for ;; HEADER SECTION
;; id = 61527
;; qr = 0	opCode: QUERY	aa = 0	tc = 0	rd = 1
;; ra = 0	ad = 0	cd = 0	rcode = NoError
;; qdCount = 1	anCount = 0	nsCount = 0	arCount = 0

;; QUESTION SECTION (1 record):
;; zpzmfhGQfHxeVuel.home.lan    IN      A       
:Net::DNS::Packet
Did you mean?  encode_tlv
[-] Call stack:
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/proto/dns/resolver.rb:136:in `send'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/net/dns/resolver.rb:894:in `query'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/exploit/remote/dns/client.rb:61:in `query'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/exploit/remote/dns/enumeration.rb:118:in `dns_get_a'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/exploit/remote/dns/enumeration.rb:366:in `dns_wildcard_enabled?'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/gather/enum_dns.rb:71:in `run'
[*] Auxiliary module execution completed
```

Reply:
Had a chance to look at this a bit more thorougly. The lines just above that were hidden by GitHub make it clear that this code path is intended to handle messages from both `Net::DNS` and `Dnsruby`. Looks like this change should be merged into that block making it:

```ruby
      case argument
      when Dnsruby::Message
        packet = argument
        packet_data = packet.encode
      when Net::DNS::Packet, Resolv::DNS::Message
        packet = Rex::Proto::DNS::Packet.encode_drb(argument)
        packet_data = packet.data
      else
        packet = make_query_packet(argument,type,cls)
        packet_data = packet.data
      end

      # Store packet_data for performance improvements,
      # so methods don't keep on calling Packet#data
      packet_size = packet_data.size
```

[make_query_packet](https://github.com/bluemonk/net-dns/blob/21d32c0b2e54ccdd3f9f27c53fe518316bf719f9/lib/net/dns/resolver.rb) is provided by Net::DNS and should return a Net::DNS object, meaning `#data` is the appropriate method to use in this case.

Reply:
This change also fixed the issue I was running into with the `enum_dns` module. I did have to work around another, unrelated issue by setting `DnsNote` to false though.

Reply:
Fixed the unrelated `enum_dns` issue in #16347 that I noticed while testing these changes to ensure that that was no regression.

Reply:
In the second `when` statement, the `encode_drb` returns a `Dnsruby::Message` object, so I think `encode` will be correct there. I think the only issue there is in the third one, which, as you say, returns a `Net::DNS::Packet` object.